common/src/chain/transaction/signature/sighash/hashable.rs

```diff
@@ -0,0 +1,242 @@
+        &self,
+        stream: &mut DefaultHashAlgoStream,
+        mode: sighashtype::SigHashType,
+        _target_input: &TxInput,
+        target_input_num: usize,
+    ) -> Result<(), TransactionSigError> {
+        match mode.outputs_mode() {
+            sighashtype::OutputsMode::All => {
+                hash_encoded_to(self, stream);
+            }
+            sighashtype::OutputsMode::None => (),
+            sighashtype::OutputsMode::Single => {
+                let output = self.get(target_input_num).ok_or({
+                    TransactionSigError::InvalidInputIndex(target_input_num, self.len())
+                })?;
+                hash_encoded_to(&output, stream);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SignatureHashableInputs<'a> {
+    inputs: &'a [TxInput],
+    /// Include utxos of the inputs to make it possible to verify the inputs scripts and amounts without downloading the full transactions
+    inputs_utxos: &'a [&'a TxOutput],
+}
+
+impl<'a> SignatureHashableInputs<'a> {
+    pub fn new(
+        inputs: &'a [TxInput],
+        inputs_utxos: &'a [&'a TxOutput],
+    ) -> Result<Self, TransactionSigError> {
+        if inputs.len() != inputs_utxos.len() {
+            return Err(TransactionSigError::InvalidUtxoCountVsInputs(
+                inputs_utxos.len(),
+                inputs.len(),
+            ));
+        }
+
+        let result = Self {
+            inputs,
+            inputs_utxos,
+        };
+
+        Ok(result)
+    }
+}
+
+impl SignatureHashableElement for SignatureHashableInputs<'_> {
+    fn signature_hash(
+        &self,
+        stream: &mut DefaultHashAlgoStream,
+        mode: sighashtype::SigHashType,
+        target_input: &TxInput,
+        target_input_num: usize,
+    ) -> Result<(), TransactionSigError> {
+        if target_input_num >= self.inputs.len() {
+            return Err(TransactionSigError::InvalidInputIndex(
+                target_input_num,
+                self.inputs.len(),
+            ));
+        }
+
+        match mode.inputs_mode() {
+            sighashtype::InputsMode::CommitWhoPays => {
+                {
+                    // Commit inputs
+                    let inputs = self.inputs;
+                    hash_encoded_to(&(inputs.len() as u32), stream);
+                    for input in inputs {
+                        hash_encoded_to(&input.outpoint(), stream);
+                    }
+                }
+
+                {
+                    // Commit inputs' utxos
+                    let inputs_utxos = self.inputs_utxos;
+                    hash_encoded_to(&(inputs_utxos.len() as u32), stream);
+                    for input_utxo in inputs_utxos {
+                        hash_encoded_to(&input_utxo, stream);
+                    }
+                }
+            }
+            sighashtype::InputsMode::AnyoneCanPay => {
+                hash_encoded_to(&target_input.outpoint(), stream);
+            }
```

Comment:
Not adding the spent UTXO in this case?

Reply:
Good question. I'm still not sure about this, to be honest. The main complaint about this problem (that required adding the utxo) was nullified in segwit with an attack that made it possible to manipulate the user due to the presence of two outputs with similar amounts (and possibly destinations). So the question is, would it even help here? I don't know.

Reply:
Seems like bitcoin does it, so I'll do it too. Nothing to lose here.

https://github.com/bitcoin/bitcoin/blob/a0d37d1d230b837ebb0ec8f609885ab89280936c/src/script/interpreter.cpp#L1534